### PR TITLE
image-rs: add image block device dm-verity and mount

### PR DESCRIPTION
**Pull request by ChengyuZhu6**
Implement image block device integrity check using dm-verity and mount the verity device in image-rs.
Notes: the kernel should be upgrade to >=6.4 for mounting the verity device with erofs tarfs.

### CODE REVIEW

**Comment by ChengyuZhu6** [comment]
> ` the kernel should be upgrade to >=6.4 for mounting the verity device with erofs.` This is not accurate, only erofs tarfs support was landed in Linux 6.4. Otherwise, if you just mount exist erofs images much like ext4 with dm-verity, there is no such requirement (5.4, 5.10, 5.15, 6.1 LTS are all supported. )

Thanks, I have updated.

File: image-rs/src/image.rs [hunk 1]
```diff
@@ -39,7 +41,6 @@ use crate::nydus::{service, utils};
 /// The reason for using the `/run` directory here is that in general HW-TEE,
 /// the `/run` directory is mounted in `tmpfs`, which is located in the encrypted memory protected by HW-TEE.
 pub const IMAGE_SECURITY_CONFIG_DIR: &str = "/run/image-security";
-
```
Comment (jiangliu) [inline image-rs/src/image.rs:42]: Unrelated change?

File: image-rs/src/image.rs [hunk 2]
```diff
@@ -421,6 +422,33 @@ impl ImageClient {
 
         Ok(image_id)
     }
+    /// set_image creates a mapping backed by image block device <source_device_path> and
```
Comment (jiangliu) [inline image-rs/src/image.rs:425]: Add a blank line above?

File: image-rs/src/verity.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    Ok(decoded_data.hash)
+}
+/// Calculates and permanently stores hash verification data for data_device.
+/// It will return the root hash.
+pub fn create_hash_device(
+    verity_options: &mut DmVerityOption,
+    data_device_path: &str,
+) -> Result<String> {
+    let veritysetup_path: &str = veritysetup_exists()?;
+    let output = Command::new(veritysetup_path)
```
Comment (jiangliu) [inline image-rs/src/verity.rs:88]: How about using https://docs.rs/devicemapper/latest/devicemapper/struct.DM.html to get rid of the dependency on veritysetup binary?

Comment (ChengyuZhu6) [reply image-rs/src/verity.rs:88]: > I haven't used devicemapper, but @jiangliu's comment seems right to me. Related to this: it looks like `verity_option` doesn't get sanity-checked or sanitized anywhere, so its values get plugged directly into the args to `versitysetup` when issuing the `Command`. Using a library call (@jiangliu's suggestion) and/or sanity-checking those values might make error handling/checking easier.

Yes, I agree with @jiangliu's comment and I am working on using the devicemapper library to handle/check the options.

File: image-rs/src/verity.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
+use anyhow::Result;
```
Comment (jiangliu) [inline image-rs/src/verity.rs:1]: How about adding a copyright header?

File: image-rs/Cargo.toml [hunk 5]
```diff
@@ -24,6 +25,7 @@ hex = { version = "0.4.3", optional = true }
 lazy_static = { version = "1.4.0", optional = true }
 libc = "0.2"
 log = "0.4.14"
+loopdev ="0.1.0"
```
Comment (jiangliu) [inline image-rs/Cargo.toml:28]: `loopdev = "0.4.0"`?

File: image-rs/src/verity.rs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    pub hash: String,
+}
+
+/// Creates a mapping with <name> backed by data_device <source_device_path>
+/// and using hash_device for in-kernel verification.
+/// It will return the verity block device Path "/dev/mapper/<name>"
+/// Notes: the data device and the hash device are the same one.
+pub fn create_verity_device(
+    verity_option: &DmVerityOption,
+    source_device_path: &str,
```
Comment (jiangliu) [inline image-rs/src/verity.rs:29]: How about using `source_device_path: &Path` here and `source_device_path.display()` below?
So the caller will get simpler.

File: image-rs/src/verity.rs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+) -> Result<String> {
+    let dm = DM::new()?;
+    let verity_name = DmName::new(&verity_option.hash)?;
+    let id = DevId::Name(verity_name);
+    let opts = DmOptions::default().set_flags(DmFlags::DM_READONLY);
+    let device_num_blocks = verity_option.blocknum.parse::<u64>().unwrap();
```
Comment (jiangliu) [inline image-rs/src/verity.rs:35]: Please return error instead of `unwrap()` here.

Comment (jiangliu) [reply image-rs/src/verity.rs:35]: The data source is untrusted and unverified, so we can't unwrap() here to avoid possible attack.

File: image-rs/src/verity.rs [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+use devicemapper::{DevId, DmFlags, DmName, DmOptions, DM};
+use serde::{Deserialize, Serialize};
+use serde_json;
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct DmVerityOption {
+    /// Hash algorithm for dm-verity.
+    pub hashtype: String,
+    /// Used block size for the data device.
+    pub blocksize: String,
```
Comment (jiangliu) [inline image-rs/src/verity.rs:12]: how about using `u64` instead of `String`?

File: image-rs/src/verity.rs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+) -> Result<String> {
+    let dm = DM::new()?;
+    let verity_name = DmName::new(&verity_option.hash)?;
+    let id = DevId::Name(verity_name);
+    let opts = DmOptions::default().set_flags(DmFlags::DM_READONLY);
+    let device_num_blocks = verity_option.blocknum.parse::<u64>().unwrap();
+    let device_block_size = verity_option.blocksize.parse::<u64>().unwrap();
+    let hash_block_size = verity_option.hashsize.parse::<u64>().unwrap();
+    let hash_start_block: u64 =
+        (device_num_blocks * device_block_size + hash_block_size - 1) / hash_block_size;
```
Comment (jiangliu) [inline image-rs/src/verity.rs:39]: Please help to verify:
- device_block_size and hash_block_size is power of two.
- hash_block_size is not zero

Comment (jiangliu) [reply image-rs/src/verity.rs:39]: BTW, we have `verity_option.offset` to calculate `hash_start_block`?

File: image-rs/src/verity.rs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+pub fn get_verity_device_name(verity_options: &str) -> Result<String> {
+    let parsed_data = decode_verity_options(verity_options)?;
+    Ok(parsed_data.hash)
+}
+
+pub fn decode_verity_options(verity_options: &str) -> Result<DmVerityOption> {
+    let decoded = base64::decode(verity_options)?;
+    let parsed_data = serde_json::from_slice::<DmVerityOption>(&decoded)?;
```
Comment (jiangliu) [inline image-rs/src/verity.rs:90]: Please help to validate parsed DmVerityOption here.

File: image-rs/src/verity.rs (hunk 9 above)
Comment (ChengyuZhu6) [reply image-rs/src/verity.rs:39]: Thanks, I'll update it.

File: image-rs/src/verity.rs (hunk 10 above)
Comment (ChengyuZhu6) [reply image-rs/src/verity.rs:90]: Thanks, I'll update it.

File: image-rs/src/verity.rs (hunk 8 above)
Comment (ChengyuZhu6) [reply image-rs/src/verity.rs:12]: Thanks, I'll update it.

File: image-rs/src/verity.rs (hunk 7 above)
Comment (ChengyuZhu6) [reply image-rs/src/verity.rs:35]: Thanks, I'll update it.

File: image-rs/src/verity.rs (hunk 6 above)
Comment (ChengyuZhu6) [reply image-rs/src/verity.rs:29]: Thanks, I'll update it.

File: image-rs/src/verity.rs (hunk 4 above)
Comment (ChengyuZhu6) [reply image-rs/src/verity.rs:1]: Thanks, I'll add a copyright.

File: image-rs/src/image.rs [hunk 11]
```diff
@@ -146,7 +148,33 @@ impl Default for ImageClient {
         }
     }
 }
-
+/// set_image creates a mapping backed by image block device <source_device_path> and
+/// decoding <verity_options> for in-kernel verification. And mount the verity device
+/// to <mount_path> with <mount_type>.
+/// It will return the verity device path if succeeds and return an error if fails .
+pub fn mount_image_block_with_integrity(
```
Comment (jiangliu) [inline image-rs/src/image.rs:155]: Do we need the counterpart `unmount_image_block_with_integrity()` to clean up all resources?
